src/sentry/storage/service.py

```diff
@@ -0,0 +1,174 @@
+        return self._make_client({"organization": organization_id, "project": project_id})
+
+
+class StorageClient:
+    def __init__(self, usecase: str, claim: dict, pool: HTTPConnectionPool, jwt_secret: str):
+        self.pool = pool
+        self.jwt_secret = jwt_secret
+        self.usecase = usecase
+        self.claim = claim
+
+    def _make_headers(self, permission: Permission) -> dict:
+        now = int(timezone.now().timestamp())
+        exp = now + JWT_VALIDITY
+        claims = {
+            "iat": now,
+            "exp": exp,
+            **self.claim,
+            "permissions": [permission],
+        }
+
+        authorization = jwt.encode(claims, self.jwt_secret)
+        return {"Authorization": authorization}
+
+    def put(
+        self,
+        contents: bytes | IO[bytes],
+        id: str | None = None,
+        compression: Compression | None = None,
+    ) -> str:
+        """
+        Uploads the given `contents` to blob storage.
+
+        If no `id` is provided, one will be automatically generated and returned
+        from this function.
+
+        The storage service will manage storing the contents compressed, unless
+        the `compression` parameter is set, in which case the contents are treated
+        as already compressed, and no double-compression will happen.
+
+        The `"uncompressible"` compression is special in the sense that it denotes
+        a file format that is already internally compressed, and does not benefit
+        from another general-purpose compression algorithm.
+        """
+        headers = self._make_headers("write")
+        body = BytesIO(contents) if isinstance(contents, bytes) else contents
+        original_body: IO[bytes] = body
+
+        if not compression:
+            cctx = zstandard.ZstdCompressor()
+            body = cctx.stream_reader(original_body)
+            headers["Content-Encoding"] = "zstd"
+        elif compression != "uncompressible":
+            headers["Content-Encoding"] = compression
+
+        compression_used = headers.get("Content-Encoding", "none")
+        with measure_storage_put(None, self.usecase, compression_used) as measurement:
+            response = self.pool.request(
+                "PUT",
+                f"/{id}" if id else "/",
+                body=body,
+                headers=headers,
+                preload_content=True,
+                decode_content=True,
+            ).json()
+
+            measurement.upload_size = body.tell()
+            if not compression:
+                metrics.distribution(
+                    "storage.put.size",
+                    original_body.tell(),
+                    tags={"usecase": self.usecase, "compression": "none"},
+                    unit="byte",
+                )
+
+            return response["key"]
+
+    def get(
+        self, id: str, accept_compression: Sequence[Compression] | None = None
+    ) -> tuple[IO[bytes], Compression | None]:
```

Comment:
_nit_: What if this becomes a named tuple right away, too? That way we can more easily extend it without breaking users that might destructure like `payload, compression = client.get(...)` and we also don't encourage patterns like `payload = client.get(...)[0]`.